src/languages/latex.js

```diff
@@ -1,70 +1,243 @@
 /*
 Language: LaTeX
-Author: Vladimir Moskva <vladmos@gmail.com>
+Author: Benedikt Wilde <bwilde@posteo.de>
 Website: https://www.latex-project.org
 Category: markup
 */
 
+import * as regex from '../lib/regex.js';
+
+/** @type LanguageFn */
 export default function(hljs) {
-  var COMMAND = {
-    className: 'tag',
+  const KNOWN_CONTROL_WORDS = regex.either(...[
+      '(?:NeedsTeXFormat|RequirePackage|GetIdInfo)',
+      'Provides(?:Expl)?(?:Package|Class|File)',
+      '(?:DeclareOption|ProcessOptions)',
+      '(?:documentclass|usepackage|input|include)',
+      'makeat(?:letter|other)',
+      'ExplSyntax(?:On|Off)',
+      '(?:new|renew|provide)?command',
+      '(?:re)newenvironment',
+      '(?:New|Renew|Provide|Declare)(?:Expandable)?DocumentCommand',
+      '(?:New|Renew|Provide|Declare)DocumentEnvironment',
+      '(?:(?:e|g|x)?def|let)',
+      '(?:begin|end)',
+      '(?:part|chapter|(?:sub){0,2}section|(?:sub)?paragraph)',
+      'caption',
+      '(?:label|(?:eq|page|name)?ref|(?:paren|foot|super)?cite)',
+      '(?:alpha|beta|[Gg]amma|[Dd]elta|(?:var)?epsilon|zeta|eta|[Tt]heta|vartheta)',
+      '(?:iota|(?:var)?kappa|[Ll]ambda|mu|nu|[Xx]i|[Pp]i|varpi|(?:var)rho)',
+      '(?:[Ss]igma|varsigma|tau|[Uu]psilon|[Pp]hi|varphi|chi|[Pp]si|[Oo]mega)',
+      '(?:frac|sum|prod|lim|infty|times|sqrt|leq|geq|left|right|middle|[bB]igg?)',
+      '(?:[lr]angle|q?quad|[lcvdi]?dots|d?dot|hat|tilde|bar)'
+    ].map(word => word + '(?![a-zA-Z@:_])'));
+  const L3_REGEX = new RegExp([
+      '(?:__)?[a-zA-Z]{2,}_[a-zA-Z](?:_?[a-zA-Z])+:[a-zA-Z]*',
+      '[lgc]__?[a-zA-Z]+_[a-zA-Z](?:_?[a-zA-Z])*[a-zA-Z]',
+      '[qs]__?[a-zA-Z](?:_?[a-zA-Z])+',
+      'use(?:_i)?:[a-zA-Z]*',
+      '(?:else|fi|or):',
+      '(?:if|cs|exp):w',
+      '(?:hbox|vbox):n',
+      '::[a-zA-Z]_unbraced',
+      '::[a-zA-Z:]'
+    ].map(pattern => pattern + '(?![a-zA-Z:_])').join('|'));
+  const L2_VARIANTS = [
+    {begin: /[a-zA-Z@]+/}, // control word
+    {begin: /[^a-zA-Z@]?/} // control symbol
+  ];
+  const DOUBLE_CARET_VARIANTS = [
+    {begin: /\^{6}[0-9a-f]{6}/},
+    {begin: /\^{5}[0-9a-f]{5}/},
+    {begin: /\^{4}[0-9a-f]{4}/},
+    {begin: /\^{3}[0-9a-f]{3}/},
+    {begin: /\^{2}[0-9a-f]{2}/},
+    {begin: /\^{2}[\u0000-\u007f]/}
```

Comment:
Is this truly meant to also include the control characters below ASCII 32?

Reply:
Yes. Honestly, I don't know how to input most of these (or which can even be input in a meaningful way), but some could definitely occur.

Some background on what this notation actually does in TeX: When you input `^^` followed by any ASCII character (with a number `n` between 0 and 127), this sequence acts like a single character. If `n` is below `64`, it acts like the character `n+64`, otherwise it acts like the character `n-64`. This way one can input character that would otherwise be hard to input, i.e. those control characters. For example `^^J` will result in a newline character and `^^` followed by a newline would result in a `J`.

Obviously, the more useful use of this notation is the first one, i.e. inputting control characters through letters and symbols. I just didn't see any reason to restrict the real rule (which is, all ASCII characters work). Is there any?

Reply:
I suppose not, just my linter flagged it as weird (and it is a little). :-)  So it was worth asking.